src/service.py

```diff
@@ -49,15 +52,25 @@ def configure(self, snap_config: dict[str, Any]) -> None:
         self.snap_client.set(snap_config, typed=True)
 
 
-def snap_install(resource: str) -> SnapService:
-    """Install or refresh the snap, and return the snap service.
+def snap_install(resource: Optional[str]) -> SnapService:
+    """Install the snap, and return the snap service.
+
+    Before installing the snap, it will remove the canonical and the upstream snap to ensure that
+    the snap installed is not a resource. Removing the golang exporter also ensure that there is
+    no conflict using the one from Canonical.
 
     Raises an exception on error.
     """
+    snap.remove([SNAP_NAME, UPSTREAM_SNAP])
     try:
-        logger.debug("installing snap.")
-        logger.debug("fetching from resource.")
-        snap_client = snap.install_local(resource, dangerous=True)
+        logger.debug("installing %s from snapcraft store", SNAP_NAME)
+        if resource:
```

Comment:
If we're dropping support for resource, can we remove this logic altogether?  I heard there was a limitation related to charmhub about removing resource metadata from a charm, but it's not confirmed.  What happens if we remove the resource from charm metadata?

If we cannot remove the resource from the charm metadata now that the charm is published, then I'm in favour of retaining support.  Otherwise it's confusing if the charm has a resource that isn't used.

Reply:
My gut feeling is that it's not possible like when you add a charm config, but I didn't find documentation about this. I've asked in the charm development room in matrix and let's see.

If it's possible I agree with you

Reply:
> like when you add a charm config

Oh it's not possible to remove charm config too?  Do you know of docs that cover this?  That doesn't make sense.

Reply:
Ok, it sounds like it's fine to remove resources or config.  The only issue is that it can make some things non-backward compatible (eg. if users were using a custom snap as a resource, now they can't)